File: src-model/model/celestials/parts/Ring.java
package model.celestials.parts;

import model.celestials.CelestialBody;
import model.interfaces.Drawable;
import model.type.DrawableResolution;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.jogamp.opengl.util.texture.Texture;

/**
 * Object which stores the ring parameters for {@link Sphere} part of {@link CelestialBody} entity.
 */
public class Ring implements Drawable {

    private String textureFile;
    private double innerRadius;
    private double outerRadius;

    /* ========== PUBLIC ========== */
    @JsonCreator
    public Ring(String textureFile, double innerRadius, double outerRadius) {
        this.textureFile = textureFile;
        this.innerRadius = innerRadius;
        this.outerRadius = outerRadius;

        this.resolution = DrawableResolution.VERY_FAR;
    }

    /* ========== PRIVATE ========== */
    @JsonIgnore
    private Texture texture;
    @JsonIgnore
    private DrawableResolution resolution;

    /* ========== PROPERTIES ========== */
    public String getTextureFile() {
        return textureFile;
    }

    public void setTextureFile(String textureFile) {
        this.textureFile = textureFile;
    }

    public double getInnerRadius() {
        return innerRadius;
    }

    public void setInnerRadius(double innerRadius) {
        this.innerRadius = innerRadius;
    }

    public double getOuterRadius() {
        return outerRadius;
    }

    public void setOuterRadius(double outerRadius) {
        this.outerRadius = outerRadius;
    }

    public Texture getTexture() {
        return texture;
    }

    public void setTexture(Texture texture) {
        this.texture = texture;
    }

    public DrawableResolution getResolution() {
        return resolution;
    }

    public void setResolution(DrawableResolution resolution) {
        this.resolution = resolution;
    }
}

File: src-model/model/celestials/parts/Ring.kt
package model.celestials.parts

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.jogamp.opengl.util.texture.Texture
import engine.components.drawers.DrawableRing
import engine.math.Radius
import engine.utils.TextureLoader
import model.celestials.CelestialBody
import model.interfaces.Drawable
import model.type.DrawableResolution
import javax.media.opengl.GL2

/**
 * Object which stores the ring parameters for [Sphere] part of [CelestialBody] entity.
 */
open class Ring
@JsonCreator
constructor(
        var textureFile: String?,
        var innerRadius: Radius,
        var outerRadius: Radius
) : Drawable {
    @JsonIgnore
    var resolution: DrawableResolution = DrawableResolution.VERY_FAR

    @JsonIgnore
    var drawable : DrawableRing? = null

    override fun draw(gl: GL2) {
        if (this.drawable == null) {
            this.drawable = DrawableRing(this, TextureLoader.getTexture(gl, textureFile))
        }

        this.drawable?.draw(gl)
    }
}